kpatch-build/create-diff-object.c

```diff
@@ -146,13 +152,78 @@ static int kpatch_mangled_strcmp(char *s1, char *s2)
 
 static int rela_equal(struct rela *rela1, struct rela *rela2)
 {
+#ifdef __powerpc__
+	struct rela *rela3, *rela4;
+	struct section *toc1, *toc2;
+	struct kpatch_elf *elf1, *elf2;
+#endif
+
 	if (rela1->type != rela2->type ||
 	    rela1->offset != rela2->offset)
 		return 0;
 
 	if (rela1->string)
 		return rela2->string && !strcmp(rela1->string, rela2->string);
 
+#ifdef __powerpc__
+/*
+ * Relocation section '.rela.toc' at offset 0x91f0 contains 122 entries:
+ *     Offset             Info             Type               Symbol's Value  Symbol's Name + Addend
+ *     [...]
+ *     0000000000000090  0000001c00000026 R_PPC64_ADDR64         0000000000000000 .rodata.__dev_remove_pack.str1.8 + 0
+ *     0000000000000098  0000001a00000026 R_PPC64_ADDR64         0000000000000000 .rodata.__dev_getfirstbyhwtype.str1.8 + 0
+ *     00000000000000a0  0000001a00000026 R_PPC64_ADDR64         0000000000000000 .rodata.__dev_getfirstbyhwtype.str1.8 + 10
+ *     00000000000000a8  000000cb00000026 R_PPC64_ADDR64         0000000000000000 dev_base_lock + 0
+ *
+ * Relocation section '.rela.text.netdev_master_upper_dev_get' at offset 0xe38 contains 10 entries:
+ *     Offset             Info             Type               Symbol's Value  Symbol's Name + Addend
+ *     [...]
+ *     00000000000000a0  0000004e00000032 R_PPC64_TOC16_HA       0000000000000000 .toc + 98
+ *     00000000000000a8  0000004e00000040 R_PPC64_TOC16_LO_DS    0000000000000000 .toc + 98
+ *     00000000000000ac  0000004e00000032 R_PPC64_TOC16_HA       0000000000000000 .toc + a0
+ *     00000000000000b0  0000004e00000040 R_PPC64_TOC16_LO_DS    0000000000000000 .toc + a0
+ *     00000000000000b4  000000b90000000a R_PPC64_REL24          0000000000000000 printk + 0
+ *     00000000000000bc  000000a10000000a R_PPC64_REL24          0000000000000000 dump_stack + 0
+ *
+ * with -mcmodel=large on ppc64le, gcc might generate entries in .toc section for symbol
+ * reference in relocation sections. There are chances of symbol referenced by toc + offset
+ * may point to different symbols in original and patched .o. Compare the symbols by
+ * reading its corresponding toc + offset rela from .toc section.
+ */
+	if (!strncmp(rela1->sym->name, ".toc", 4) &&
+	    !strncmp(rela2->sym->name, ".toc", 4)) {
+
+		elf1 = container_of(&rela1->sym->sec->list, struct kpatch_elf, sections);
+		toc1 = find_section_by_name(&elf1->sections, ".rela.toc");
```

Comment:
Also maybe rename `toc1` to `toc_relasec1` or something?  It's longer, but it much better describes what it is.